src/main.rs

```diff
@@ -71,46 +82,102 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     // Initialize jsonrpc server
     let (jsonrpc_tx, jsonrpc_rx) = mpsc::unbounded_channel::<PortalEndpoint>();
+    let jsonrpc_trin_config = trin_config.clone();
     let web3_server_task = tokio::task::spawn_blocking(|| {
-        launch_jsonrpc_server(trin_config, infura_project_id, jsonrpc_tx);
+        launch_jsonrpc_server(jsonrpc_trin_config, infura_project_id, jsonrpc_tx);
     });
 
-    tokio::spawn(async move {
-        info!(
-            "About to spawn portal p2p with boot nodes: {:?}",
-            portalnet_config.bootnode_enrs
-        );
-
-        let (mut p2p, events) = PortalnetProtocol::new(discovery, portalnet_config)
-            .await
-            .unwrap();
+    // Setup Overlay database
+    let db = Arc::new(setup_overlay_db(discovery.read().await.local_enr()));
 
+    // Spawn main JsonRpc Handler
+    let jsonrpc_discovery = discovery.clone();
+    tokio::spawn(async move {
         let rpc_handler = JsonRpcHandler {
-            discovery: p2p.discovery.clone(),
+            discovery: jsonrpc_discovery,
             jsonrpc_rx,
             state_tx,
             history_tx,
         };
 
-        tokio::spawn(events.process_discv5_requests());
         tokio::spawn(rpc_handler.process_jsonrpc_requests());
+
         if let Some(handler) = state_handler {
             tokio::spawn(handler.handle_client_queries());
         }
         if let Some(handler) = history_handler {
             tokio::spawn(handler.handle_client_queries());
         }
+    });
 
-        // hacky test: make sure we establish a session with the boot node
-        p2p.ping_bootnodes().await.unwrap();
-
-        tokio::signal::ctrl_c()
-            .await
-            .expect("failed to pause until ctrl-c");
-    })
-    .await
-    .unwrap();
+    debug!("Selected networks to spawn: {:?}", trin_config.networks);
+
+    for network in trin_config.networks.iter() {
+        match network.as_str() {
+            STATE_NETWORK => {
+                let discovery_state = discovery.clone();
```

Comment:
Sorry for being so nitpicky - but while I do think these are an improvement over `discovery_1`. It's still pretty confusing imo. I understand the implication of this naming ("**discovery** for the **state** network") - but if I was brand new reading this, I'd be pretty confused as I'd interpret it more like this is the "current state of discovery", or  this is the database state for `db_state`,  and below is the "history of discovery". IDK, maybe I'm being a little paranoid, but I always err on the side of readability vs. small names. Something like `state_subnet_discovery` is quite explicit.  I do prefer explicit and obvious names, especially when we're working on a project with many different contributors jumping in and out of the codebase.

I'm definitely open to hear pushback though.

Reply:
Thinking about this now, we can also make the arc cloning more readable and explicit, for example, instead of `discovery.clone()`, we can say `Arc::clone(&discovery)`, how do you feel about this? Reference: [link](https://doc.rust-lang.org/std/sync/struct.Arc.html#cloning-references)

For the naming, probably `state_network_discovery` makes more sense?

Reply:
:+1: to both of those!